bugtool/dump/bpfmap.go

```diff
@@ -0,0 +1,108 @@
+package dump
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"path"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/maps/ipcache"
+	"github.com/cilium/cilium/pkg/maps/lxcmap"
+	"github.com/cilium/cilium/pkg/maps/tunnel"
+	"github.com/cilium/ebpf"
+)
+
+// This provides a lookup from map name to a factory to create
+// bpfmap Tasks.
+//
+// TODO: Provide generic fallback for non registered types?
+var typedBPFMapRegistry = map[string]taskFactory{
+	"ipcache":     bpfMapFactory[ipcache.Key, ipcache.RemoteEndpointInfo]{},
+	"call_policy": bpfMapFactory[lxcmap.EndpointKey, lxcmap.EndpointInfo]{},
+	"lxc":         bpfMapFactory[lxcmap.EndpointKey, lxcmap.EndpointInfo]{},
+	"tunnel_map":  bpfMapFactory[tunnel.TunnelKey, tunnel.TunnelValue]{},
+}
+
+func getTaskFactory(mapName string) (taskFactory, error) {
+	n := strings.TrimPrefix(mapName, "cilium_")
+	if tf, ok := typedBPFMapRegistry[n]; ok {
+		return tf, nil
+	}
+	return nil, fmt.Errorf("no known bpfmap map %q", mapName)
+}
+
+// NewBPFMap constructs a new BPF map to dump a pinned bpf map file.
+func NewPinnedBPFMap[KT fmt.Stringer, VT any](pinnedFile string) *BPFMap[KT, VT] {
+	return &BPFMap[KT, VT]{
+		base: base{
+			Name: path.Base(pinnedFile),
+			Kind: "BPFMap",
+		},
+		PinnedFile: pinnedFile,
+	}
+}
+
+// BPFMap is a generic type that implements the Task interface
+// by dumping bpf maps.
+type BPFMap[KT fmt.Stringer, VT any] struct {
```

Comment:
This will definitely prevent backporting because older branches don't build Cilium with a version of Go that supports generics